packages/itwinui-react/src/core/Table/Table.tsx

```diff
@@ -578,18 +580,47 @@ export const Table = <
     );
   }, [data, getSubRows]);
 
-  const [rowHasParent] = React.useState(() => new WeakSet<T>());
-
   const getSubRowsWithSubComponents = React.useCallback(
-    (originalRow: T) => {
-      if (!rowHasParent.has(originalRow)) {
-        rowHasParent.add(originalRow);
-        return [originalRow];
+    (originalRow: T, relativeIndex: number) => {
+      console.log('row', originalRow);
+      // If originalRow represents a sub-component, don't add any sub-rows to it
+      if (originalRow[iuiId as any]) {
+        return [];
+      }
+
+      // If originalRow represents the top-most level row, add itself as a sub-row that will represent a subComponent.
+      // Add a symbol as a key on this sub-row to distinguish it from the real row.
+      // This distinction is needed as the sub-row needs to have all fields of the row since react-table expects even sub-rows to be rows (Row<T>).
+      if (originalRow.subRows) {
+        return originalRow.subRows as T[];
       }
 
-      return (originalRow.subRows as T[]) || [];
+      return [
+        {
+          [iuiId]: `subcomponent-${relativeIndex}`,
+          ...originalRow,
+        },
+      ];
+    },
+    [],
+  );
+
+  /**
+   * Gives `subComponent` a unique id since `subComponent` has the same react-table id as its parent row. Avoiding duplicate react-table ids prevents react-table errors.
+   */
+  const getRowIdWithSubComponents = React.useCallback(
+    (originalRow: T, relativeIndex: number, parent?: Row<T>) => {
+      const mainRowId =
+        getRowId?.(originalRow, relativeIndex, parent) ?? `${relativeIndex}`;
+
+      // If the row contains the Symbol, it indicates that the current row is a sub-component row. We need to append the ID passed by user with its according sub-component ID.
+      return originalRow[iuiId as any]
+        ? `${mainRowId}-${
+            parent ? `${parent.id}.${relativeIndex}` : relativeIndex
+          }`
+        : mainRowId;
```

Comment:
Avoid nested ternaries. This can be made more readable by using an `if` condition.

```suggestion
      const mainRowId =
        getRowId?.(originalRow, relativeIndex, parent) ?? `${relativeIndex}`;

      // If the row contains the Symbol, it indicates that the current row is a sub-component row.
      // We need to append the ID passed by user with its according sub-component ID.
      if (originalRow[iuiId as any]) {
        const suffix = parent ? `${parent.id}.${relativeIndex}` : relativeIndex;
        return `${mainRowId}-${suffix}`;
      }

      return mainRowId;
```

Reply:
Actually I'm not sure that this function is even how it should be. Looking at the [docs](https://react-table-v7-docs.netlify.app/docs/api/useTable#table-options), the default value is: `(row, relativeIndex, parent) => parent ? [parent.id, relativeIndex].join('.') : relativeIndex`, whereas this function falls back to just `relativeIndex`.

```suggestion
      const defaultRowId = parent
        ? [parent.id, relativeIndex].join('.')
        : relativeIndex;

      const mainRowId = getRowId?.(originalRow, relativeIndex, parent);

      // If the row contains the Symbol, it indicates that the current row is a sub-component row.
      // We need to append the ID passed by user with its according sub-component ID.
      if (mainRowId && originalRow[iuiId as any]) {
        return `${mainRowId}-${defaultRowId}`;
      }

      return mainRowId ?? defaultRowId;
```

Reply:
Not sure if I understand it correctly but this might cause an error due to race condition since `getRowId` takes precedence before the sub-rows are generated. It seems that it doesn't gets inside the if statement since the id passed by user is defined while the sub-rows are not yet generated, which ends up in a loop.
![image](https://github.com/user-attachments/assets/cdf3c703-1080-4ed5-91d9-c27662f2de52)
 
Additionally, since the original return type is `string`, and the current return type in this function is `string | number`, I think I also have to make `defaultRowId` a string so it won't error out.

Can I do something like this?

````jsx
const getRowIdWithSubComponents = React.useCallback(
    (originalRow: T, relativeIndex: number, parent?: Row<T>) => {
      const defaultRowId = `${
        parent ? [parent.id, relativeIndex].join('.') : relativeIndex
      }`;
      const mainRowId = getRowId?.(originalRow, relativeIndex, parent);
      // If the row contains the Symbol, it indicates that the current row is a sub-component row.
      // We need to append the ID passed by user with its according sub-component ID.
      if (originalRow[iuiId as any]) {
        return mainRowId ? `${mainRowId}-${defaultRowId}` : defaultRowId;
      }
      return mainRowId ?? defaultRowId;
    },
    [getRowId],
  );
````

Reply:
I'm not sure I understand your comment about race conditions. Doesn't it fall back to `defaultRowId` in all cases?

Reply:
I tried printing out `rowIdFromUser` and the original rows from `getSubRowsWithSubComponent` and saw that it kept returning `rowIdFromUser`, not the expected fallback value, before the sub-rows got to be created.
![image](https://github.com/user-attachments/assets/4bbe4f53-0031-488b-bbbf-1dfec501ffa2)

Reply:
I wonder if this is a problem with `0` being a falsy value. Based on your ternary condition, the equivalent fallback return value would need to use `||` instead of `??`.

<details><summary>Full code</summary>

```tsx
const defaultRowId = parent
  ? `${parent.id}.${relativeIndex}`
  : `${relativeIndex}`;
const rowIdFromUser = getRowId?.(originalRow, relativeIndex, parent);
// If the row contains the Symbol, it indicates that the current row is a sub-component row.
// We need to append the ID passed by user with its according sub-component ID.
if (rowIdFromUser && originalRow[iuiId as any]) {
  return `${rowIdFromUser}-${defaultRowId}`;
}
return rowIdFromUser || defaultRowId;
```

</details> 

This does raise the question: is `0` a valid value that we want to use, or do we want to discard it? If it's valid, then we should check using `??` and `!== undefined` instead of loose boolean checks.

Reply:
I used `rowIdFromUser !== undefined` for the if condition, put `??` back to the final return and it behaved correctly. Hmm, I think it should be kept since it's the index of the first row (?).